src/storage/src/hummock/iterator/concat_inner.rs

```diff
@@ -102,16 +103,15 @@ impl<TI: SSTableIteratorType> HummockIterator for ConcatIteratorInner<TI> {
     async fn seek(&mut self, key: &[u8]) -> HummockResult<()> {
         let table_idx = self
             .tables
-            .partition_point(|table| match TI::DIRECTION {
-                FORWARD => {
+            .partition_point(|table| match Self::Direction::direction() {
+                DirectionEnum::Forward => {
```

Comment:
Can we ensure that this `match` can be resolved in compile-time?

Reply:
May add `const fn` on `direction()` function.

Reply:
Seems the function in the trait cannot be made `const`. We may fall back to const generic. You may need something like this @wenym1 

https://github.com/singularity-data/risingwave/blob/4f31c18e99cf0bfe87f958047b8f8aa59dbf9cce/src/stream/src/executor/hash_join.rs#L35-L46

Reply:
I have added `#[inline(always)]` in the `direction()`. Can this possibly be resolved in compile-time?

Reply:
Not sure about that...

Reply:
Just run a simple bench about this discussion. It seems that dispatching using enum returned from the function call, regardless of whether it is `inline` or not, can get slightly better performance than using const dispatch.

Bench output is as followed.
```
const                   time:   [507.05 ps 508.36 ps 509.67 ps]                   

enum not inline         time:   [502.60 ps 503.98 ps 505.55 ps]                             

enum inline             time:   [504.40 ps 505.58 ps 506.67 ps]     
```

Code is attached below.
```
use criterion::{black_box, Criterion, criterion_group, criterion_main};

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);

enum Num {
    One,
    Two,
}

trait GetEnum {
    fn get_num() -> Num;
}

struct GetEnumNotInline;

impl GetEnum for GetEnumNotInline {
    fn get_num() -> Num {
        Num::One
    }
}

struct GetNumInline;

impl GetEnum for GetNumInline {
    #[inline(always)]
    fn get_num() -> Num {
        Num::One
    }
}

fn const_dispatch<const NUM: usize>() {
    let mut sum = 0;
    for _ in 0..black_box(1000000) {
        match NUM {
            1 => sum += 1,
            2 => sum += 2,
            _ => unreachable!(),
        }
    }
    assert_eq!(sum, 1000000);
}

fn enum_dispatch<T: GetEnum>() {
    let mut sum = 0;
    for _ in 0..black_box(1000000) {
        match T::get_num() {
            Num::One => sum += 1,
            Num::Two => sum += 2,
        }
    }
    assert_eq!(sum, 1000000);
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("const", |b| b.iter(const_dispatch::<1>));

    c.bench_function("enum not inline", |b| b.iter(enum_dispatch::<GetEnumNotInline>));

    c.bench_function("enum inline", |b| b.iter(enum_dispatch::<GetNumInline>));
}
```

Reply:
Seems okay 🤣. https://godbolt.org/z/xnzc8Pr9z